xdsl/transforms/loop_invariant_code_motion.py

```diff
@@ -32,9 +41,66 @@ def can_be_hoisted(op: Operation, target_region: Region) -> bool | None:
     for child in op.walk():
         for operand in child.operands:
             operand_owner = operand.owner
-            assert operand_owner is not None
             if op.is_ancestor(operand_owner):
                 continue
             if target_region.is_ancestor(operand_owner):
                 return False
     return True
+
+
+def _move_loop_invariant_code(region: Region, builder: Builder):
+    # add top-level operations in the loop body to the worklist
+    worklist = deque(op for block in region.blocks for op in block.ops)
+
+    while worklist:
+        op = worklist.popleft()
+        # Skip ops that have already been moved. Check if the op can be hoisted.
+        if op.parent_region() != region:
+            continue
+
+        if not (
+            is_side_effect_free(op)
+            and is_speculatable(op)
+            and can_be_hoisted(op, region)
+        ):
+            continue
+
+        op.detach()
+        builder.insert(op)
+
+        # Since the op has been moved, we need to check its users within the
+        # top-level of the loop body.
+
+        for res in op.results:
+            for use in res.uses:
+                user = use.operation
+                if user.parent_region() == region:
+                    worklist.append(user)
+
+
+def move_loop_invariant_code(loop: scf.ForOp):
+    builder = Builder(InsertPoint.before(loop))
+    _move_loop_invariant_code(loop.body, builder)
+
+
+class LoopInvariantCodeMotion(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: scf.ForOp, rewriter: PatternRewriter) -> None:
+        move_loop_invariant_code(op)
+
+
+class LoopInvariantCodeMotionPass(ModulePass):
+    """
+    Moves operations without side effects out of loops, provided they do not depend on
+    values defined in the loops.
+    """
+
+    name = "licm"
+
+    def apply(self, ctx: Context, op: builtin.ModuleOp) -> None:
+        PatternRewriteWalker(
+            LoopInvariantCodeMotion(),
+            apply_recursively=False,
+            walk_regions_first=True,
+            walk_reverse=True,
```

Comment:
sorry if this already got answered but why do we `walk_reverse`?

Reply:
It feels like it might be the right thing to do but I couldn't come up with an example where it changes anything, so I'll leave it out.